content/post/2022-09-01-r-dependency/index.Rmd

```diff
@@ -0,0 +1,223 @@
+  unlist()
+tail(r_deps)
+```
+
+As mentioned earlier, the minimum required version can be specified with a loose or strict inequality:
+
+```{r}
+(r_deps_strict <- sum(grepl("^R \\(>\\s(.+)\\)", r_deps)))
+(r_deps_loose  <- sum(grepl("^R \\(>=\\s(.+)\\)", r_deps)))
+```
+
+You can see that using a strict inequality is indeed very uncommon (`r sprintf("%.2f%%", r_deps_strict / (r_deps_strict + r_deps_loose) * 100)` of the cases).
+
+We can now continue our analysis and extract the version number itself:
+
+```{r}
+r_deps_ver <- trimws(gsub("^R \\(>=?\\s(.+)\\)", "\\1", r_deps))
+
+r_deps_ver |> 
+  table()
+```
+
+Interestingly, you can notice that some of these version numbers don't match any actual R release.
+To confirm this, we can use the [rversions package, from R-hub](https://cran.r-project.org/package=rversions):
+
+```{r}
+setdiff(unique(r_deps_ver), rversions::r_versions()$version)
+```
+
+We can infer the reason for the mismatch for some examples in this list:
+
+-   missing `.` between version components (e.g., `2.01`, `2.50`, `3.00`, `3.60`, `4.00`)
+-   `.` replaced by `-` in the patch version number (e.g., `3.0-0`, `3.0-2`, `3.1-0`, `3.5-0`, `4.1-0`) [^3].
+-   missing patch version number (e.g., `2.0`, `2.2`, `4.3`)
+-   extra patch version number (e.g., `1.4.0`)
+-   recommended packages depend on a yet-to-be-released R version (`4.3`)
+
+[^3]: However, it is interesting to note that `package_version("3.5-0") == package_version("3.5.0")`.
+    The use of `-` instead of `.` is purely stylistic.
+
+Note that this values are not syntactically wrong, and it might in some cases be intended by the author.
+They can be read and understood by the relevant function in base R (in particular, `install.packages()`), but it is possible they do not correspond to what the package author was expecting, or trying to communicate.
+For example, in the case of `R (=> 3.60)`: even if the author really intended to depend on `R 3.6.0` as we assume here, the package cannot be installed in versions earlier than 4.0.0.
+
+To visualise the actual minimum R version corresponding to the declared R dependency, we can do the following:
+
+```{r}
+r_vers <- package_version(rversions::r_versions()$version)
+
+normalised_r_deps <- vapply(r_deps_ver, function(ver) {
+  
+  ver <- package_version(ver)
+  
+  # Here, we rely on a somewhat uncommon use of `match()`. When `match()`ing
+  # `TRUE` to something, the index of the first `TRUE` value will be returned.
+  # In other words here, we return the first R version that it superior or equal
+  # to the stated R version dependency
+  min_r_ver <- r_vers[match(TRUE, ver <= r_vers)]
+  
+  return(min_r_ver)
+  
+}, package_version("0.0.0"))
+```
+
+```{r}
+library(ggplot2)
+
+do.call(rbind, normalised_r_deps) |> 
+  as.data.frame() |> 
+  dplyr::rename(
+    major = V1,
+    minor = V2,
+    patch = V3
+  ) |> 
+  dplyr::mutate(majorminor = paste(major, minor, sep = ".")) |> 
+  ggplot(aes(y = majorminor)) +
```

Comment:
I was expecting the plot the other way round, with R versions on the x-axis (it does not mean it would be a good idea though).

Reply:
I think this is actually the recommended way to do it. Putting the R versions would:

- either look crammed
- or force me to rotate the tick labels, which is discouraged without it's difficult to read (unless you tilt your head)

Can't find a reference for this claim right now though.